### ISSUE TITLE
failing to send whispers, new twitchchat servers

### ISSUE THREAD

**Issue by itskeyes** [SReject/mTwitch#4]
Not sure if this is entirely the code's fault, but since twitch has moved their chat servers a bit, whispers are no longer sending.

**Comment by Blasman** [comment]
I have a friend who is a partnered streamer on the new Twitch servers. The mTwitch scripts were working great on the old Twitch servers. However, now that he is on the new server, the mTwitch scripts do not appear to be working at all.

**Comment by itskeyes** [comment]
Same. This bot is being ran in a large channel, and was recently moved to the new servers.

**Comment by SReject** [comment]
This should only effect the mTwitch scripts while twitch migrates to their new servers. Once migration is complete the problem should fix itself.

If it does not, I will see what I can do about getting the whisper script fixed. I will leave this report open so others can see my response :)

**Comment by itskeyes** [comment]
mIRC is reporting this: MTWITCH.CHATSERVERLISTUPDATE Unknown command

**Comment by SReject** [comment]
Are you sure you have mTwitch.Core.mrc installed?

**Comment by itskeyes** [comment]
Yes. Looks like it only happened once. Deleting above comment.

**Comment by SReject** [comment]
Adding an update here. I believe the issue arises from users that are connecting to the irc interface using the address `irc.chat.twitch.tv`. I am waiting for a pull request that should fix this issue.

**Comment by SReject** [comment]
I've updated mTwitch.Core.mrc to match `irc.chat.twitch.tv`; this should, hopefully, fix the whisper problem.

commit of fix: 11682c4ff774fafdf046d1ed3982d6af7eabc67c

**Comment by itskeyes** [comment]
I've updated mtwitch.core.mrc, and now i am receiving "[mTwitch->GroupChat] Invalid oauth token; stopping Twitch Group-Chat connection attempts."

**Comment by SReject** [comment]
I've updated `mTwitch.GroupChat.mrc` to support irc.chat.twitch.tv aswell. I'm unable to reproduce whisper issues, currently. Could those having issues list what server/address and port you are connecting to?

**Comment by itskeyes** [comment]
`{"cluster":"aws","servers":["irc.chat.twitch.tv:6667","irc.chat.twitch.tv:6667","irc.chat.twitch.tv:80","irc.chat.twitch.tv:80"],"websockets_servers":["irc-ws.chat.twitch.tv:80","irc-ws.chat.twitch.tv:80","irc-ws.chat.twitch.tv:80"]}`

That's all the ones the channel is told to use, as references by twitchs post on dev forums.

**Comment by SReject** [comment]
I've started adding debug messages throughout core and groupchat to hopefully determine where the script fails. to access the debug use `/mTwitchDebug` on **BEFORE** connecting to twitch's normal chat servers

**Comment by itskeyes** [comment]
Getting this now with the latest updates: `[mTwitch->GroupChat] Connection to Twitch Group-Chat server failed to open; retrying...`

and

```
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
[GroupChat] Connection established to ~*
[GroupChat Connect] Connection to group.tmi.twitch.tv on port 443
[GroupChat] Connection established to ~*
[GroupChat Connect] Connection to group.tmi.twitch.tv on port 443
```

**Comment by SReject** [comment]
There was a typo/logic-flaw in the previous version; I've updated to fix it

**Comment by itskeyes** [comment]
Okay, still no go. Here is the latest debug output:

```
[GroupChat Open] Connection to Twitch group-chat server failed to open; retrying...
[GroupChat Connect] Connection to group.tmi.twitch.tv on port 443
[GroupChat] Connection established to group.tmi.twitch.tv
[GroupChat Open] Registering with twitch
[$mTwitch.isServer.UpdateList] Attempting to retrieve server lists
[$mTwitch.isServer.UpdateList] Number of servers returned: 4
[$mTwitch.isServer.UpdateList] Retrieving group servers
[$mTwitch.isServer.UpdateList] Added irc.chat.twitch.tv as a general chat server
[$mTwitch.isServer.UpdateList] Added irc.chat.twitch.tv as a general chat server
[$mTwitch.isServer.UpdateList] Added irc.chat.twitch.tv as a general chat server
[$mTwitch.isServer.UpdateList] Added irc.chat.twitch.tv as a general chat server
[$mTwitch.isServer.UpdateList] Added group.tmi.twitch.tv as a group chat server
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
[$mTwitch.isServer] Called with the parameters:
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
```

**Comment by SReject** [comment]
Have updated again with even more debug messages, and fixed more flawed logic; this time in `mTwitch.GroupChat.Buffer`

**Comment by itskeyes** [comment]
Okay here we go, error logging in message groupchat parsing:

```
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
[$mTwitch.isServer] Called with the parameters:
[$mTwitch.isServer] Server found is server list as: General
[GroupChat] Captured twitch connection attempt; attempting to connect to group-chat servers
[GroupChat Connect] Attempting to connect to group.tmi.twitch.tv on port 443
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
[GroupChat] Connection to group.tmi.twitch.tv:443 established; Sending nick and oauth
[GroupChat Buffer] Adding to sendbuffer: PASS irc
[GroupChat Buffer] Adding to sendbuffer: NICK xxxxx
[GroupChat Buffer] Adding to sendbuffer: USER xxxxx ? * :Twitch User
[GroupChat Buffer] Adding to sendbuffer: CAP REQ :twitch.tv/commands twitch.tv/tags twitch.tv/membership
[GroupChat Buffer] All data sent.
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
[GroupChat Parsing] :tmi.twitch.tv NOTICE * :Error logging in
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
```

**Comment by SReject** [comment]
Q: is the "PASS irc" part edited, as that should be an oauth token

**Comment by SReject** [comment]
Updated again to fix oauth not being sent correctly

**Comment by itskeyes** [comment]
You got it dude! It's now working, here's the final debug and i've truncated and removed some parts:

```
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
[$mTwitch.isServer] Called with the parameters:
[$mTwitch.isServer] Server found is server list as: General
[GroupChat] Captured twitch connection attempt; attempting to connect to group-chat servers
[GroupChat Connect] Attempting to connect to group.tmi.twitch.tv on port 443
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
[GroupChat] Connection to group.tmi.twitch.tv:443 established; Sending nick and oauth
[GroupChat Buffer] Adding to sendbuffer: PASS oauth:<removed for security>
[GroupChat Buffer] Adding to sendbuffer: NICK <removed>
[GroupChat Buffer] Adding to sendbuffer: USER <removed> ? * :Twitch User
[GroupChat Buffer] Adding to sendbuffer: CAP REQ :twitch.tv/commands twitch.tv/tags twitch.tv/membership
[GroupChat Buffer] All data sent.
[$mTwitch.isServer] Called with the parameters:
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
[GroupChat Parsing] :tmi.twitch.tv 001 cinbot :Welcome, GLHF!
[GroupChat Parsing] :tmi.twitch.tv 002 cinbot :Your host is tmi.twitch.tv
[GroupChat Parsing] :tmi.twitch.tv 003 cinbot :This server is rather new
[GroupChat Parsing] :tmi.twitch.tv 004 cinbot :-
[GroupChat Parsing] :tmi.twitch.tv 375 cinbot :-
[GroupChat Parsing] :tmi.twitch.tv 372 cinbot :You are in a maze of twisty passages, all alike.
[GroupChat Parsing] :tmi.twitch.tv 376 cinbot :>
[GroupChat Parsing] :tmi.twitch.tv CAP * ACK :twitch.tv/commands twitch.tv/tags twitch.tv/membership
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
[$mTwitch.isServer] Called with the parameters:
[GroupChat Buffer] Adding to sendbuffer: PRIVMSG jtv :/w xxxxx <message output>
[GroupChat Buffer] All data sent.
[$mTwitch.isServer] Network matches twitch.tv or server matches twitch chat servers
```

**Comment by SReject** [comment]
Just to be clear, whispers are now working for you?

**Comment by itskeyes** [comment]
Yes. Whispers are now working again.

**Comment by SReject** [comment]
This has been fixed as of commit 00679355c71f4f15cdde5971dee28cbabdb5352d
